appserver/payara-appserver-modules/security-openid/src/main/java/fish/payara/security/openid/OpenIdAuthenticationMechanism.java

```diff
@@ -184,18 +186,18 @@ public AuthenticationStatus validateRequest(
             return authenticationController.authenticateUser(configuration, httpContext);
         }
 
-        String recievedState = request.getParameter(STATE);
+        Optional<OpenIdState> receivedState = OpenIdState.from(request.getParameter(STATE));
         String redirectURI = configuration.buildRedirectURI(request);
-        if (nonNull(recievedState)
+        if (receivedState.isPresent()
                 && request.getRequestURL().toString().equals(redirectURI)) {
 
-            OpenIdState expectedState = stateController.get(configuration, httpContext);
-            if (nonNull(expectedState)) {
-                if (expectedState.equals(recievedState)) {
+            Optional<OpenIdState> expectedState = stateController.get(configuration, httpContext);
+            if (expectedState.isPresent()) {
+                if (expectedState.get().equals(receivedState.get())) {
                     // (3) Successful Authentication Response : redirect_uri?code=abc&state=123
```

Comment:
This condition may be simplified to expectedState.equals(recievedState).

If both objects are Optional, equals of Optional delegates call to it's value. As a bonus we will get rid of get call that may throw an exception.

Reply:
Done. BTW the get() calls in this case couldn't throw an exception because of the isPresent checks.